python-lib/custom_gluon_models/arima.py

```diff
@@ -0,0 +1,134 @@
+from gluonts.model.estimator import Estimator
+from gluonts.model.forecast import SampleForecast
+from gluonts.model.predictor import RepresentablePredictor
+from gluonts.support.pandas import frequency_add
+from gluonts.core.component import validated
+from gluonts.time_feature import get_seasonality
+from custom_gluon_models.utils import cast_kwargs
+from constants import TIMESERIES_KEYS
+import pmdarima as pm
+import numpy as np
+from safe_logger import SafeLogger
+from tqdm import tqdm
+
+
+logger = SafeLogger("Forecast plugin - ARIMA")
+
+
+class ArimaPredictor(RepresentablePredictor):
+    """
+    An abstract predictor that can be subclassed by models that are not based
+    on Gluon. Subclasses should have @validated() constructors.
+    (De)serialization and value equality are all implemented on top of the
+    @validated() logic.
+
+    Parameters
+    ----------
+    prediction_length
+        Prediction horizon.
+    freq
+        Frequency of the predicted data.
+    """
+
+    # TODO implement custom serializer
+
+    @validated()
+    def __init__(self, prediction_length, freq, trained_models, lead_time=0):
+        super().__init__(freq=freq, lead_time=lead_time, prediction_length=prediction_length)
+        self.trained_models = trained_models
+
+    def predict(self, dataset, **kwargs):
+        """
+
+        Args:
+            dataset (gluonts.dataset.common.Dataset): Dataset after wich to predict forecasts.
+
+        Yields:
+            SampleForecast of predictions.
+        """
+        logger.info("Prediction timeseries ...")
+        for i, item in tqdm(enumerate(dataset)):
+            yield self.predict_item(item, self.trained_models[i])
+
+    def predict_item(self, item, trained_model):
+        """Compute quantiles using the confidence intervals of auto_arima.
+
+        Args:
+            item (DataEntry): One timeseries.
+            trained_model (list): List of trained auto_arima models.
+
+        Returns:
+            SampleForecast of quantiles.
+        """
+        start_date = frequency_add(item["start"], len(item["target"]))
+
+        prediction_external_features = self._set_prediction_external_features(item)
+
+        samples = []
+        for alpha in np.arange(0.02, 1.01, 0.02):
```

Comment:
Why is `alpha` hardcoded? Shouldn't we pickup the settings chosen by the user in the Predict recipe?

Reply:
I hardcoded alpha to generate 99 samples that are all the percentiles from 1 to 99.
A GluonTS model outputs during prediction a SampleForecast class that contains many prediction samples generated by the model. These samples are then used to compute quantiles that I retrieve later based on the user input.

Reply:
Okay! If this is to allow the computation of quantile-based performance metrics, that's fine. 

However, how can we ensure the correct alpha is computed when the user chooses a confidence interval in the predict recipe?

Reply:
when a user chooses a confidence interval, I convert it into 2 quantiles (90% confidence interval becomes quantiles 0.05 and 0.95), then I call `sample_forecasts.quantile_ts(quantile)` for each of the 2 quantiles, where `sample_forecasts` is basically the output of `predictor.predict(gluon_dataset)`. So the 2 quantiles retrieved from  `sample_forecasts` are quantiles I set before with the alpha.

Reply:
Ok super, closing this then